### PR TITLE
controllers: refactor code and start informers first, fix template/render race

### PR DESCRIPTION
**Pull request by runcom**
This patch does various things, all related:

- start the informers for controllers before running them to follow what
https://github.com/kubernetes/sample-controller/blob/master/main.go#L64-L71
does and to avoid races already spotted in unit tests
(#457)
- move the clients builder code from cmd/common to a new package just for that
under the new internal/ folder so nobody but us can use that
- move common controllers code under pkg/controller/common to be reused
- create an e2e only clientset, avoiding us to type client version
everytime (eg client.CoreV1()..). This is a need cause if in the future
the api change, we don't play grep for a week replacing old apis...
- fix template/render controllers race

Closes: #338
Closes: #385

Signed-off-by: Antonio Murdaca <runcom@linux.com>

### CODE REVIEW

**Comment by cgwalters** [comment]
I think with stuff like this it'd help me a lot to have a "reference operator" codebase.  There's [the kube sample controller](https://github.com/kubernetes/sample-controller/) but our code is different enough I'm not sure if it's doing the same thing.

**Comment by runcom** [comment]
> I think with stuff like this it'd help me a lot to have a "reference operator" codebase. There's [the kube sample controller](https://github.com/kubernetes/sample-controller/) but our code is different enough I'm not sure if it's doing the same thing.

cool, and from that example operator it looks like informers are indeed started before the controller itself https://github.com/kubernetes/sample-controller/blob/master/main.go#L64-L71 - I've rebased this as well

/assign @LorbusChris

**Comment by runcom** [comment]
This has been updated and does fix #338 as well (by waiting on the template controller :tada:  ):

```
I0317 16:27:21.495296       1 node_controller.go:383] Pool master is unconfigured, pausing 5s for renderer to initialize
I0317 16:27:21.495462       1 node_controller.go:383] Pool worker is unconfigured, pausing 5s for renderer to initialize
I0317 16:27:21.495537       1 render_controller.go:380] Error syncing machineconfigpool master: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.495584       1 render_controller.go:380] Error syncing machineconfigpool worker: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.501140       1 render_controller.go:380] Error syncing machineconfigpool master: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.501180       1 render_controller.go:380] Error syncing machineconfigpool worker: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.511368       1 render_controller.go:380] Error syncing machineconfigpool master: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.511407       1 render_controller.go:380] Error syncing machineconfigpool worker: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.531622       1 render_controller.go:380] Error syncing machineconfigpool master: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.531647       1 render_controller.go:380] Error syncing machineconfigpool worker: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.571897       1 render_controller.go:380] Error syncing machineconfigpool worker: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.571897       1 render_controller.go:380] Error syncing machineconfigpool master: ControllerConfig has not completed: completed(false) running(false) failing(true)
I0317 16:27:21.671461       1 render_controller.go:500] Generated machineconfig worker-b058168e9ec42a1704343e84bb2dc7eb from 4 configs: [{MachineConfig  00-worker  machineconfiguration.openshift.io/v1  } {MachineConfig  00-worker-ssh  machineconfiguration.openshift.io/v1  } {MachineConfig  01-worker-container-runtime  machineconfiguration.openshift.io/v1  } {MachineConfig  01-worker-kubelet  machineconfiguration.openshift.io/v1  }]
I0317 16:27:21.684235       1 render_controller.go:500] Generated machineconfig master-34908aca2b6075843cc77cd59449a0df from 4 configs: [{MachineConfig  00-master  machineconfiguration.openshift.io/v1  } {MachineConfig  00-master-ssh  machineconfiguration.openshift.io/v1  } {MachineConfig  01-master-container-runtime  machineconfiguration.openshift.io/v1  } {MachineConfig  01-master-kubelet  machineconfiguration.openshift.io/v1  }]
```

**Comment by runcom** [comment]
@cgwalters @kikisdeliveryservice @LorbusChris ptal

**Comment by runcom** [comment]
Weird flakes... Looks like the wrong git reference is being tested (that line in mcd_test.go isn't even there in current revision).

/retest

**Comment by kikisdeliveryservice** [comment]
LGTM! Really like some of the changes here, will leave it to @cgwalters / @LorbusChris to give it one more set of :eyes:

File: pkg/controller/render/render_controller.go [hunk 1]
```diff
@@ -425,6 +440,35 @@ func (ctrl *Controller) syncMachineConfigPool(key string) error {
...
+		return fmt.Errorf("controllerConfigList is empty")
+	}
+
+	cur, err := ctrl.ccLister.Get(cc[0].GetName())
+	if err != nil {
+		return err
+	}
+
+	if cur.Generation != cur.Status.ObservedGeneration {
+		return fmt.Errorf("status for ControllerConfig %s is being reported for %d, expecting it for %d", cc[0].GetName(), cur.Status.ObservedGeneration, cur.Generation)
```
Comment (cgwalters) [inline pkg/controller/render/render_controller.go]: Feels like this isn't an *error* but eh.

**Pull request merged**